### PR TITLE
refactor: Turn linearity checking into separate compiler stage

### LINKED ISSUE #272: Turn linearity check into separate compiler stage

### PR DESCRIPTION
**Pull request by mark-koch**
Closes #272

### CODE REVIEW

File: tests/error/linear_errors/break_unused.py [hunk 1]
```diff
@@ -26,7 +26,7 @@ def foo(i: int) -> bool:
         if i == 0:
             break
         i -= 1
-        b ^= measure(q)
+        b &= measure(q)
```
Comment (mark-koch) [inline tests/error/linear_errors/break_unused.py:29]: for some reason `bool.__xor__` is missing (see #274).

The test worked before since the linearity violation triggered an error before we got to this point. Now we type check everything first, so this has to be fixed (similar for the remaining modified tests)

File: guppylang/cfg/builder.py [hunk 2]
```diff
@@ -270,7 +271,9 @@ def visit_NamedExpr(self, node: ast.NamedExpr) -> ast.Name:
         # assignment statement and replace the expression with `x`.
         if not isinstance(node.target, ast.Name):
             raise InternalGuppyError(f"Unexpected assign target: {node.target}")
-        assign = ast.Assign(targets=[node.target], value=self.visit(node.value))
+        assign = ast.Assign(
+            targets=[copy.deepcopy(node.target)], value=self.visit(node.value)
+        )
```
Comment (aborgna-q) [inline guppylang/cfg/builder.py:276]: Is this a drive-by?

**Review by aborgna-q — APPROVED** [review]
Nice and tidy.

The comments on `_check_comprehension` are quite useful.

File: guppylang/cfg/builder.py (hunk 2 above)
Comment (mark-koch) [reply guppylang/cfg/builder.py:276]: No this is required now that we annotate the LHS of assignments with types

https://github.com/CQCL/guppylang/blob/f0dec41f8e722cbbbf2a9f98a7980a85f67c9a56/guppylang/checker/stmt_checker.py#L49-L56

If we don't make a copy, the type checker will see that a type has been assigned when the variable is used the next time and won't bother checking again which leads incorrect behaviour
